File: app/GameTables.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Web.Script.Serialization;

namespace BakhmutMap {
    public class GameRow {
        public double distance { get; set; }
        public double elevation { get; set; }
        public double seconds { get; set; }
        public string source { get; set; }
    }
    public class GameTable {
        public string id { get; set; }
        public string shell { get; set; }
        public int rings { get; set; }
        public string source { get; set; }
        public List<GameRow> rows { get; set; }
        public string weapon { get; set; }
        public string trajectory { get; set; }
        public int unitsPerCircle { get; set; }
        public GameTable() {weapon="mortar82";trajectory="high";unitsPerCircle=6000;}
    }
    public class GameSolution {
        public bool Available { get; set; }
        public string Message { get; set; }
        public double Distance { get; set; }
        public double AzimuthDegrees { get; set; }
        public double Elevation { get; set; }
        public double Seconds { get; set; }
        public double LowerDistance { get; set; }
        public double UpperDistance { get; set; }
        public int UnitsPerCircle {get;set;}
        public string Source {get;set;}
        public GameSolution() {UnitsPerCircle=6000;}
        public int AzimuthUnits { get { return (int)Math.Round(AzimuthDegrees*UnitsPerCircle/360,MidpointRounding.AwayFromZero)%UnitsPerCircle; } }
        public double ElevationDegrees {get{return ElevationUnits*360.0/UnitsPerCircle;}}
        public int ElevationUnits { get { return (int)Math.Round(Elevation,MidpointRounding.AwayFromZero); } }
    }
    // Interpolates only the supplied game lookup tables. No physical ballistic model.
    public static class GameTables {
        private static bool Finite(double x) { return !Double.IsNaN(x)&&!Double.IsInfinity(x); }
        public static List<GameTable> Load(string path) {
            List<GameTable> tables=new JavaScriptSerializer().Deserialize<List<GameTable>>(File.ReadAllText(path));
            if(tables==null||tables.Count==0)throw new FormatException("No game tables");
            HashSet<string> ids=new HashSet<string>();
            foreach(GameTable table in tables) {
                if(table==null||String.IsNullOrWhiteSpace(table.id)||!ids.Add(table.id)||String.IsNullOrWhiteSpace(table.shell)||String.IsNullOrWhiteSpace(table.source)||table.rows==null||table.rows.Count<2)throw new FormatException("Invalid game table");
                if(table.weapon=="mortar82") {if(table.unitsPerCircle!=6000||table.rings<0||table.rings>4||table.trajectory!="high")throw new FormatException("Invalid mortar profile");}
                else if(table.weapon=="m777") {if(table.unitsPerCircle!=6400||table.rings<1||table.rings>5||(table.trajectory!="high"&&table.trajectory!="low"))throw new FormatException("Invalid M777 profile");}
                else throw new FormatException("Unknown weapon profile");
                double previous=-1;
                foreach(GameRow row in table.rows) {
                    if(row==null||!Finite(row.distance)||!Finite(row.elevation)||!Finite(row.seconds)||row.distance<=previous||row.distance<=0||row.elevation<=0||row.elevation>=table.unitsPerCircle/4||row.seconds<=0)throw new FormatException("Invalid game table row: "+table.id);
                    previous=row.distance;
                }
            }
            return tables;
        }
        public static GameSolution Calculate(GameTable table,MapPoint position,MapPoint target) {
            GameSolution result=new GameSolution();
            if(table!=null)result.UnitsPerCircle=table.unitsPerCircle;
            if(!Coordinates.Valid(position)||!Coordinates.Valid(target)) { result.Message="Выберите позицию и цель на карте.";return result; }
            result.Distance=Coordinates.Distance(position,target);
            if(result.Distance<0.001) { result.Message="Позиция и цель совпадают; азимут не определён.";return result; }
            result.AzimuthDegrees=(Math.Atan2(target.X-position.X,target.Z-position.Z)*180/Math.PI+360)%360;
            if(table==null) { result.Message="Выберите игровую таблицу.";return result; }
            GameRow first=table.rows.First(),last=table.rows.Last();
            if(result.Distance<first.distance-1e-7||result.Distance>last.distance+1e-7) {
                result.Message="Вне выбранной таблицы: "+first.distance+"–"+last.distance+" м. Измените заряд, траекторию или позицию.";return result;
            }
            GameRow lower=first,upper=last;
            double sampleDistance=Math.Max(first.distance,Math.Min(last.distance,result.Distance));
            foreach(GameRow row in table.rows) {
                if(row.distance<=sampleDistance)lower=row;
                if(row.distance>=sampleDistance) {upper=row;break;}
            }
            double ratio=upper.distance==lower.distance?0:(sampleDistance-lower.distance)/(upper.distance-lower.distance);
            result.Elevation=lower.elevation+(upper.elevation-lower.elevation)*ratio;
            result.Seconds=lower.seconds+(upper.seconds-lower.seconds)*ratio;
            result.LowerDistance=lower.distance;result.UpperDistance=upper.distance;
            result.Source=String.Join(" | ",new[]{lower.source??table.source,upper.source??table.source}.Distinct());
            result.Available=true;return result;
        }
    }
}

File: web/domain/game-tables.ts
import { distance, validPoint, type MapPoint } from './coordinates.ts';

export type WeaponId = 'mortar82' | 'm777';
export type Trajectory = 'high' | 'low';
export interface GameRow { distance: number; elevation: number; seconds: number; source?: string | null }
export interface GameTable {
  id: string; shell: string; rings: number; source: string; rows: GameRow[];
  weapon: WeaponId; trajectory: Trajectory; unitsPerCircle: 6000 | 6400;
}
export type SolutionReason = 'missing-points' | 'coincident-points' | 'missing-table' | 'out-of-range' | null;
export interface GameSolution {
  Available: boolean; Message: string | null; Reason: SolutionReason;
  Distance: number; AzimuthDegrees: number; AzimuthUnits: number;
  Elevation: number; ElevationUnits: number; ElevationDegrees: number;
  Seconds: number; LowerDistance: number; UpperDistance: number;
  UnitsPerCircle: number; Source: string | null;
}

function object(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}
function text(value: unknown): value is string { return typeof value === 'string' && value.trim().length > 0; }

/** Validate untrusted JSON and apply the constructor defaults of GameTable in C#. */
export function parseGameTables(value: unknown): GameTable[] {
  if (!Array.isArray(value) || value.length === 0) throw new Error('No game tables');
  const ids = new Set<string>();
  return value.map((item: unknown) => {
    if (!object(item) || !text(item.id) || ids.has(item.id) || !text(item.shell) || !text(item.source) || !Array.isArray(item.rows) || item.rows.length < 2) throw new Error('Invalid game table');
    ids.add(item.id);
    const id = item.id;
    const weapon = item.weapon === undefined ? 'mortar82' : item.weapon;
    const trajectory = item.trajectory === undefined ? 'high' : item.trajectory;
    const units = item.unitsPerCircle === undefined ? 6000 : item.unitsPerCircle;
    const rings = item.rings === undefined ? 0 : item.rings;
    if (typeof rings !== 'number' || !Number.isInteger(rings)) throw new Error('Invalid charge');
    if (weapon === 'mortar82') {
      if (units !== 6000 || rings < 0 || rings > 4 || trajectory !== 'high') throw new Error('Invalid mortar profile');
    } else if (weapon === 'm777') {
      if (units !== 6400 || rings < 1 || rings > 5 || (trajectory !== 'high' && trajectory !== 'low')) throw new Error('Invalid M777 profile');
    } else throw new Error('Unknown weapon profile');
    let previous = -1;
    const rows = item.rows.map((row: unknown): GameRow => {
      if (!object(row) || typeof row.distance !== 'number' || !Number.isFinite(row.distance) || row.distance <= previous || row.distance <= 0
        || typeof row.elevation !== 'number' || !Number.isFinite(row.elevation) || row.elevation <= 0 || row.elevation >= (units as number) / 4
        || typeof row.seconds !== 'number' || !Number.isFinite(row.seconds) || row.seconds <= 0
        || (row.source != null && typeof row.source !== 'string')) throw new Error(`Invalid game table row: ${id}`);
      previous = row.distance;
      return { distance: row.distance, elevation: row.elevation, seconds: row.seconds, source: row.source as string | null | undefined };
    });
    return { id: item.id, shell: item.shell, source: item.source, rings, weapon, trajectory: trajectory as Trajectory, unitsPerCircle: units as 6000 | 6400, rows };
  });
}

/** Reproduce the existing LuckyUI M116 game profiles without changing base tables. */
export function createLuckyTables(mortarData: unknown, m777Data: unknown): GameTable[] {
  const base = [...parseGameTables(mortarData), ...parseGameTables(m777Data)];
  const smoke = base.filter((t) => t.weapon === 'm777' && t.shell === 'M107 HE').map((t) => ({
    ...t, id: `${t.id}-smoke`, shell: 'M116 SMOKE', rows: t.rows.map((row) => ({ ...row })),
  }));
  const all = [...base, ...smoke];
  if (new Set(all.map((t) => t.id)).size !== all.length) throw new Error('Duplicate game profile ID');
  return all;
}

export function roundAwayFromZero(value: number): number {
  return Math.sign(value) * Math.floor(Math.abs(value) + 0.5);
}

/** Interpolation of supplied Arma game tables only; no new physical model. */
export function calculate(table: GameTable | null, position: MapPoint | null, target: MapPoint | null): GameSolution {
  const result: GameSolution = {
    Available: false, Message: null, Reason: null, Distance: 0, AzimuthDegrees: 0, AzimuthUnits: 0,
    Elevation: 0, ElevationUnits: 0, ElevationDegrees: 0, Seconds: 0, LowerDistance: 0, UpperDistance: 0,
    UnitsPerCircle: table?.unitsPerCircle ?? 6000, Source: null,
  };
  if (!validPoint(position) || !validPoint(target)) return { ...result, Reason: 'missing-points', Message: 'Выберите позицию и цель на карте.' };
  result.Distance = distance(position, target);
  if (result.Distance < 0.001) return { ...result, Reason: 'coincident-points', Message: 'Позиция и цель совпадают; азимут не определён.' };
  result.AzimuthDegrees = (Math.atan2(target.X - position.X, target.Z - position.Z) * 180 / Math.PI + 360) % 360;
  result.AzimuthUnits = roundAwayFromZero(result.AzimuthDegrees * result.UnitsPerCircle / 360) % result.UnitsPerCircle;
  if (table === null) return { ...result, Reason: 'missing-table', Message: 'Выберите игровую таблицу.' };
  const first = table.rows[0];
  const last = table.rows[table.rows.length - 1];
  if (result.Distance < first.distance - 1e-7 || result.Distance > last.distance + 1e-7) return {
    ...result, Reason: 'out-of-range', Message: `Вне выбранной таблицы: ${first.distance}–${last.distance} м. Измените заряд, траекторию или позицию.`,
  };
  let lower = first;
  let upper = last;
  const sampleDistance = Math.max(first.distance, Math.min(last.distance, result.Distance));
  for (const row of table.rows) {
    if (row.distance <= sampleDistance) lower = row;
    if (row.distance >= sampleDistance) { upper = row; break; }
  }
  const ratio = upper.distance === lower.distance ? 0 : (sampleDistance - lower.distance) / (upper.distance - lower.distance);
  result.Elevation = lower.elevation + (upper.elevation - lower.elevation) * ratio;
  result.Seconds = lower.seconds + (upper.seconds - lower.seconds) * ratio;
  result.LowerDistance = lower.distance;
  result.UpperDistance = upper.distance;
  result.Source = [...new Set([lower.source ?? table.source, upper.source ?? table.source])].join(' | ');
  result.ElevationUnits = roundAwayFromZero(result.Elevation);
  result.ElevationDegrees = result.ElevationUnits * 360 / result.UnitsPerCircle;
  result.Available = true;
  return result;
}
